src/main/java/picard/sam/SetNmMdAndUqTags.java

```diff
@@ -0,0 +1,113 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package picard.sam;
+
+import htsjdk.samtools.SAMException;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMFileWriter;
+import htsjdk.samtools.SAMFileWriterFactory;
+import htsjdk.samtools.SamReader;
+import htsjdk.samtools.SamReaderFactory;
+import htsjdk.samtools.reference.ReferenceSequenceFileWalker;
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.samtools.util.Log;
+import htsjdk.samtools.util.ProgressLogger;
+import picard.cmdline.CommandLineProgram;
+import picard.cmdline.CommandLineProgramProperties;
+import picard.cmdline.Option;
+import picard.cmdline.StandardOptionDefinitions;
+import picard.cmdline.programgroups.SamOrBam;
+
+import java.io.File;
+import java.util.stream.StreamSupport;
+
+/**
+ * @author Yossi Farjoun
+ */
+@CommandLineProgramProperties(
+        usage = SetNmMdAndUqTags.USAGE_SUMMARY + SetNmMdAndUqTags.USAGE_DETAILS,
+        usageShort = SetNmMdAndUqTags.USAGE_SUMMARY,
+        programGroup = SamOrBam.class
+)
+public class SetNmMdAndUqTags extends CommandLineProgram {
+    static final String USAGE_SUMMARY = "Fixes the NM, MD, and UQ tags in a SAM file.  ";
+    static final String USAGE_DETAILS = "This tool takes in a SAM or BAM file (sorted by coordinate) and calculates the NM, MD, and UQ tags by comparing with the reference."+
+            "<br />" +
+                    "This may be needed when MergeBamAlignment was run with SORT_ORDER different from 'coordinate' and thus could not fix\n"+
+                    "these tags then.<br />"+
+            "<h4>Usage example:</h4>" +
+            "<pre>" +
+            "java -jar picard.jar SetNmMDAndUqTags \\<br />" +
+            "      I=sorted.bam \\<br />" +
+            "      O=fixed.bam \\<br />"+
+            "</pre>" +
+            "<hr />";
+    @Option(doc = "The BAM or SAM file to fix.", shortName = StandardOptionDefinitions.INPUT_SHORT_NAME)
+    public File INPUT;
+
+    @Option(doc = "The fixed BAM or SAM output file. ", shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME)
+    public File OUTPUT;
+
+    @Option(doc = "Whether the file contains bisulfite sequence (used when calculating the NM tag).")
+    public boolean IS_BISULFITE_SEQUENCE = false;
+
+    @Override
+    protected String[] customCommandLineValidation() {
+        if (REFERENCE_SEQUENCE == null) {
+            return new String[]{"Must have a non-null REFERENCE_SEQUENCE"};
+        }
+        return super.customCommandLineValidation();
+    }
+
+    private final Log log = Log.getInstance(SetNmMdAndUqTags.class);
+
+    public static void main(final String[] argv) {
```

Comment:
Not sure, but I am going to leave it here as other classes have it; cleaning this up should be in all classes could be in another PR.